flow/layers/opacity_layer.cc

```diff
@@ -4,65 +4,120 @@
 
 #include "flutter/flow/layers/opacity_layer.h"
 
-#include "flutter/flow/layers/transform_layer.h"
+#include "flutter/fml/trace_event.h"
+#include "third_party/skia/include/core/SkPaint.h"
 
-namespace flutter {
+#if defined(OS_FUCHSIA)
+#include "flutter/flow/scene_update_context.h"  //nogncheck
+#endif                                          // defined(OS_FUCHSIA)
 
-OpacityLayer::OpacityLayer(int alpha, const SkPoint& offset)
-    : alpha_(alpha), offset_(offset) {}
+namespace flutter {
 
-OpacityLayer::~OpacityLayer() = default;
+constexpr float kOpacityElevationWhenUsingSystemCompositor = 0.001f;
 
-void OpacityLayer::EnsureSingleChild() {
-  FML_DCHECK(layers().size() > 0);  // OpacityLayer should never be a leaf
+#if !defined(OS_FUCHSIA)
+void OpacityLayerBase::Preroll(PrerollContext* context,
+                               const SkMatrix& matrix) {
+  const SkAlpha alpha = SkColorGetA(color());
+  const float parent_is_opaque = context->is_opaque;
 
-  if (layers().size() == 1) {
-    return;
-  }
+  context->mutators_stack.PushOpacity(alpha);
+  context->is_opaque = parent_is_opaque && (alpha == 255);
+  ContainerLayer::Preroll(context, matrix);
+  context->is_opaque = parent_is_opaque;
+  context->mutators_stack.Pop();
+}
+#endif
 
-  // Be careful: SkMatrix's default constructor doesn't initialize the matrix to
-  // identity. Hence we have to explicitly call SkMatrix::setIdentity.
-  SkMatrix identity;
-  identity.setIdentity();
-  auto new_child = std::make_shared<flutter::TransformLayer>(identity);
+OpacityLayer::OpacityLayer(SkAlpha opacity, const SkPoint& offset)
+    : OpacityLayerBase(SK_ColorBLACK,
+                       opacity,
+                       kOpacityElevationWhenUsingSystemCompositor),
+      offset_(offset) {
+  // Ensure OpacityLayer has only one direct child.
+  //
+  // This is needed to ensure that retained rendering can always be applied to
+  // save the costly saveLayer.
+  //
+  // Any children will be actually added as children of this empty
+  // ContainerLayer.
+  ContainerLayer::Add(std::make_shared<ContainerLayer>());
+}
 
-  for (auto& child : layers()) {
-    new_child->Add(child);
-  }
-  ClearChildren();
-  Add(new_child);
+void OpacityLayer::Add(std::shared_ptr<Layer> layer) {
```

Comment:
can we add an `FML_DCHECK` to ensure that this only gets called once?

Reply:
It is perfectly legal to call this method N times though; its ContainerLayer::Add that should only be called once.  I don't think I can add an FML_DCHECK to that without add a should_have_one_child() virtual, or something similar